src/box/lua/config/configdata.lua

```diff
@@ -246,6 +253,167 @@ local function effective_snapshot_dir(iconfig)
+    end
+
+    if snap_names.instance_name ~= nil and
+       snap_names.instance_name ~= config_names.instance_name then
+        error(string.format('Instance name mismatch. Snapshot: %s, ' ..
+                            'config: %s.', snap_names.instance_name,
+                            config_names.instance_name))
+    end
+
+    -- Fail early, if current UUID is not set, but no name is found
+    -- inside the snapshot file.
+    --
+    -- TODO: it's NoOp for now, as instance_uuid is automatically
+    -- generated for vshard configuration. This error and error_uuids
+    -- can't happen and consequently not tested. We must get rid of
+    -- generating UUIDs and test it properly.
+    --
+    -- For now it's impossible to recover from snaps without manually
+    -- specifying UUIDs.
+    if snap_names.instance_name == nil and
+       config_names.instance_uuid == nil then
+        error(string.format('Instance name is not set in snapshot and UUID ' ..
+                            'is missing in the config. Found %s in snapshot.',
+                            snap_names.instance_uuid))
+    end
+
+    -- Check, that every peer, which doesn't have name in snap,
+    -- has UUID in config, so that we can validate, that the right
+    -- config is applied to the instance.
+    --
+    -- Don't immediately fail on encountering the missing UUID.
+    -- Gather a batch of all instances, where UUIDs are needed and fail
+    -- only after that.
+    local error_uuids = {}
+    local missing_names = {}
+    for _, snap_peer in ipairs(snap_names.peers) do
+        if snap_peer.instance_name ~= nil then
+            goto skip
+        end
+
+        assert(snap_peer.instance_uuid ~= nil)
+        local config_peer_name = find_peer_name_by_uuid(config_names.peers,
+                                                        snap_peer.instance_uuid)
+        if config_peer_name ~= nil then
+            missing_names[config_peer_name] = snap_peer.instance_uuid
+        else
+            table.insert(error_uuids, snap_peer.instance_uuid)
+        end
+
+        ::skip::
+    end
+
+    if #error_uuids > 0 then
+        -- We don't have any info about names, which should be associated
+        -- with these UUIDs. E.g. if no UUIDs are given, master has name
+        -- set in snapshot, and replica doesn't, then master will fail with
+        -- this error, but replica - with error 'Instance name is not set'
+        -- (see above). So, replica will say, which UUID should be set,
+        -- but everyone fails to start.
+        error(string.format("Some replicas without names doesn't have UUIDs " ..
+                            "set in config. Cannot match the following UUIDs" ..
+                            " with names: %s", json.encode(error_uuids)))
+    end
+
+    -- Add replicaset_name to missing_names if needed.
+    if snap_names.replicaset_name == nil then
+        missing_names[config_names.replicaset_name] = snap_names.replicaset_uuid
+    end
```

Comment:
Nit: Is it OK to mix instance and replicaset names?

Reply:
It feels comfortable to use. I'd rather not change that

Reply:
I'll ask my question differently just in case.

We have instance names of the current instance (I), replicaset name of the current instance (R) and instance names of peers (P).

Is it right that all the code is ready for I+R and R+P name clashes?

IOW, is it right that if R == I, than adding just R into `missing_names` will not lead to unintended effects for the instance name? Will not removing of R from the `missing_names` miss I? The same for R == P (for some peer).